codec-http/src/main/java/io/netty/handler/codec/http/HttpRequestCompressor.java

```diff
@@ -0,0 +1,263 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.codec.MessageToByteEncoder;
+import io.netty.handler.codec.compression.Brotli;
+import io.netty.handler.codec.compression.BrotliEncoder;
+import io.netty.handler.codec.compression.BrotliOptions;
+import io.netty.handler.codec.compression.CompressionOptions;
+import io.netty.handler.codec.compression.DeflateOptions;
+import io.netty.handler.codec.compression.GzipOptions;
+import io.netty.handler.codec.compression.SnappyFrameEncoder;
+import io.netty.handler.codec.compression.StandardCompressionOptions;
+import io.netty.handler.codec.compression.ZlibCodecFactory;
+import io.netty.handler.codec.compression.ZlibWrapper;
+import io.netty.handler.codec.compression.Zstd;
+import io.netty.handler.codec.compression.ZstdEncoder;
+import io.netty.handler.codec.compression.ZstdOptions;
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.StringUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * A {@code ChannelOutboundHandler} that encodes (compresses) the body of HTTP-Requests using the given encoding.
+ * <p>
+ * <b>Supported encodings:</b>
+ * <ul>
+ * <li>gzip (default)</li>
+ * <li>deflate</li>
+ * <li>snappy</li>
+ * <li>br (depends on "com.aayushatharva.brotli4j:brotli4j")</li>
+ * <li>zstd (depends on "com.github.luben:zstd-jni")r</li>
+ * </ul>
+ * <p>
+ * <b>Note for zstd:</b>
+ * <a href="https://github.com/netty/netty/issues/15340">
+ * you should define a threshold that is greater or equal to the configured block size</a>.
+ * <p>
+ * <b>How-To Use</b>
+ * <p>
+ * For each new connection, add the handler to the end of the user pipeline.
+ * <pre>
+ * HttpClient
+ *   .create()
+ *   // add the handler
+ *   .observe((con, newState) -> {
+ *     // use CONFIGURED for HTTP/1.1 connections
+ *     // use STREAM_CONFIGURED for HTTP/2 connections
+ *     if (newState == HttpClientState.CONFIGURED || newState == HttpClientState.STREAM_CONFIGURED) {
+ *       con.addHandlerLast(new HttpRequestCompressor("gzip"));
+ *     }
+ *   })
+ *   .build();
```

Comment:
Why Reactor Netty example?

Reply:
Is Reactory Netty not a valid example? And, to be honest, I have never used Netty as an HTTP client.

Reply:
This is Netty, so it's good to have Netty-related examples only. Pipeline configuration would be nice.